File: csharp/EngineCore/Light.cs
using System;

namespace vkEngine.EngineCore
{
    public unsafe abstract class Light
    {
        private readonly UInt32 entity;

        protected Light(UInt32 entity)
        {
            this.entity = entity;
        }

        public UInt32 Entity => entity;

        internal static void RegisterNativeFunctions(NativeFunctions* functions)
        {
            DirectionalLight.RegisterDirectionalLightNativeFunctions(functions);
            PointLight.RegisterPointLightNativeFunctions(functions);
            SpotLight.RegisterSpotLightNativeFunctions(functions);
        }
    }

    public unsafe sealed class DirectionalLight : Light
    {
        private static delegate* unmanaged[Cdecl]<UInt32, NVec3*, void> getColor;
        private static delegate* unmanaged[Cdecl]<UInt32, NVec3*, void> setColor;
        private static delegate* unmanaged[Cdecl]<UInt32, float> getIntensity;
        private static delegate* unmanaged[Cdecl]<UInt32, float, void> setIntensity;

        public DirectionalLight(UInt32 entity) : base(entity) { }

        internal static void RegisterDirectionalLightNativeFunctions(NativeFunctions* functions)
        {
            getColor = functions->GetDirectionalLightColor;
            setColor = functions->SetDirectionalLightColor;
            getIntensity = functions->GetDirectionalLightIntensity;
            setIntensity = functions->SetDirectionalLightIntensity;
        }

        public NVec3 Color
        {
            get
            {
                NVec3 value = default;
                getColor(Entity, &value);
                return value;
            }
            set
            {
                NVec3 nativeValue = value;
                setColor(Entity, &nativeValue);
            }
        }

        public float Intensity
        {
            get => getIntensity(Entity);
            set => setIntensity(Entity, value);
        }
    }

    public unsafe sealed class PointLight : Light
    {
        private static delegate* unmanaged[Cdecl]<UInt32, NVec3*, void> getColor;
        private static delegate* unmanaged[Cdecl]<UInt32, NVec3*, void> setColor;
        private static delegate* unmanaged[Cdecl]<UInt32, float> getIntensity;
        private static delegate* unmanaged[Cdecl]<UInt32, float, void> setIntensity;
        private static delegate* unmanaged[Cdecl]<UInt32, float> getRadius;
        private static delegate* unmanaged[Cdecl]<UInt32, float, void> setRadius;

        public PointLight(UInt32 entity) : base(entity) { }

        internal static void RegisterPointLightNativeFunctions(NativeFunctions* functions)
        {
            getColor = functions->GetPointLightColor;
            setColor = functions->SetPointLightColor;
            getIntensity = functions->GetPointLightIntensity;
            setIntensity = functions->SetPointLightIntensity;
            getRadius = functions->GetPointLightRadius;
            setRadius = functions->SetPointLightRadius;
        }

        public NVec3 Color
        {
            get
            {
                NVec3 value = default;
                getColor(Entity, &value);
                return value;
            }
            set
            {
                NVec3 nativeValue = value;
                setColor(Entity, &nativeValue);
            }
        }

        public float Intensity
        {
            get => getIntensity(Entity);
            set => setIntensity(Entity, value);
        }

        public float Radius
        {
            get => getRadius(Entity);
            set => setRadius(Entity, value);
        }
    }

    public unsafe sealed class SpotLight : Light
    {
        private static delegate* unmanaged[Cdecl]<UInt32, NVec3*, void> getColor;
        private static delegate* unmanaged[Cdecl]<UInt32, NVec3*, void> setColor;
        private static delegate* unmanaged[Cdecl]<UInt32, float> getIntensity;
        private static delegate* unmanaged[Cdecl]<UInt32, float, void> setIntensity;
        private static delegate* unmanaged[Cdecl]<UInt32, float> getRadius;
        private static delegate* unmanaged[Cdecl]<UInt32, float, void> setRadius;
        private static delegate* unmanaged[Cdecl]<UInt32, float> getInnerCone;
        private static delegate* unmanaged[Cdecl]<UInt32, float, void> setInnerCone;
        private static delegate* unmanaged[Cdecl]<UInt32, float> getOuterCone;
        private static delegate* unmanaged[Cdecl]<UInt32, float, void> setOuterCone;

        public SpotLight(UInt32 entity) : base(entity) { }

        internal static void RegisterSpotLightNativeFunctions(NativeFunctions* functions)
        {
            getColor = functions->GetSpotLightColor;
            setColor = functions->SetSpotLightColor;
            getIntensity = functions->GetSpotLightIntensity;
            setIntensity = functions->SetSpotLightIntensity;
            getRadius = functions->GetSpotLightRadius;
            setRadius = functions->SetSpotLightRadius;
            getInnerCone = functions->GetSpotLightInnerCone;
            setInnerCone = functions->SetSpotLightInnerCone;
            getOuterCone = functions->GetSpotLightOuterCone;
            setOuterCone = functions->SetSpotLightOuterCone;
        }

        public NVec3 Color
        {
            get
            {
                NVec3 value = default;
                getColor(Entity, &value);
                return value;
            }
            set
            {
                NVec3 nativeValue = value;
                setColor(Entity, &nativeValue);
            }
        }

        public float Intensity
        {
            get => getIntensity(Entity);
            set => setIntensity(Entity, value);
        }

        public float Radius
        {
            get => getRadius(Entity);
            set => setRadius(Entity, value);
        }

        public float InnerCone
        {
            get => getInnerCone(Entity);
            set => setInnerCone(Entity, value);
        }

        public float OuterCone
        {
            get => getOuterCone(Entity);
            set => setOuterCone(Entity, value);
        }
    }
}

File: src/interop/light.cpp
#include <interop/light.hpp>
#include <render/render.hpp>
#include <scene.hpp>
#include <glm/common.hpp>
#include <glm/trigonometric.hpp>

namespace vke_interop
{
    static inline entt::entity ToEntity(uint32_t entity)
    {
        return static_cast<entt::entity>(entity);
    }

    template <vke_render::AllowedLightType T>
    static T &GetLightWithoutCheckByEntity(uint32_t entity)
    {
        return vke_render::Renderer::GetInstance()->lightManager->GetLightWithoutCheckByEntity<T>(ToEntity(entity));
    }

    template <vke_render::AllowedLightType T>
    static void MarkLightDirty()
    {
        vke_render::Renderer::GetInstance()->lightManager->MarkDirty<T>();
    }

    static void UpdateSpotLight(uint32_t entity, vke_render::SpotLight &light)
    {
        auto *lightManager = vke_render::Renderer::GetInstance()->lightManager.get();
        lightManager->MarkDirty<vke_render::SpotLight>();
        if (light.CastShadow())
            lightManager->UpdateSpotShadow(ToEntity(entity));
    }

    static Vector3<float> ToInteropColor(const glm::vec4 &colorWithIntensity)
    {
        return Vector3<float>{colorWithIntensity.x, colorWithIntensity.y, colorWithIntensity.z};
    }

    static glm::vec3 ToGlm(const Vector3<float> &value)
    {
        return glm::vec3(value.x, value.y, value.z);
    }

    template <vke_render::AllowedLightType T>
    static void GetLightColor(uint32_t entity, Vector3<float> *color)
    {
        *color = ToInteropColor(GetLightWithoutCheckByEntity<T>(entity).colorWithIntensity);
    }

    template <vke_render::AllowedLightType T>
    static void SetLightColor(uint32_t entity, const Vector3<float> *color)
    {
        auto &light = GetLightWithoutCheckByEntity<T>(entity);
        glm::vec3 rgb = ToGlm(*color);
        light.colorWithIntensity = glm::vec4(rgb, light.colorWithIntensity.w);
        MarkLightDirty<T>();
    }

    template <vke_render::AllowedLightType T>
    static float GetLightIntensity(uint32_t entity)
    {
        return GetLightWithoutCheckByEntity<T>(entity).colorWithIntensity.w;
    }

    template <vke_render::AllowedLightType T>
    static void SetLightIntensity(uint32_t entity, float intensity)
    {
        GetLightWithoutCheckByEntity<T>(entity).colorWithIntensity.w = intensity;
        MarkLightDirty<T>();
    }

    void VKE_INTEROP_CDECL GetDirectionalLightColor(uint32_t entity, Vector3<float> *color)
    {
        GetLightColor<vke_render::DirectionalLight>(entity, color);
    }

    void VKE_INTEROP_CDECL SetDirectionalLightColor(uint32_t entity, const Vector3<float> *color)
    {
        SetLightColor<vke_render::DirectionalLight>(entity, color);
    }

    float VKE_INTEROP_CDECL GetDirectionalLightIntensity(uint32_t entity)
    {
        return GetLightIntensity<vke_render::DirectionalLight>(entity);
    }

    void VKE_INTEROP_CDECL SetDirectionalLightIntensity(uint32_t entity, float intensity)
    {
        SetLightIntensity<vke_render::DirectionalLight>(entity, intensity);
    }

    void VKE_INTEROP_CDECL GetPointLightColor(uint32_t entity, Vector3<float> *color)
    {
        GetLightColor<vke_render::PointLight>(entity, color);
    }

    void VKE_INTEROP_CDECL SetPointLightColor(uint32_t entity, const Vector3<float> *color)
    {
        SetLightColor<vke_render::PointLight>(entity, color);
    }

    float VKE_INTEROP_CDECL GetPointLightIntensity(uint32_t entity)
    {
        return GetLightIntensity<vke_render::PointLight>(entity);
    }

    void VKE_INTEROP_CDECL SetPointLightIntensity(uint32_t entity, float intensity)
    {
        SetLightIntensity<vke_render::PointLight>(entity, intensity);
    }

    float VKE_INTEROP_CDECL GetPointLightRadius(uint32_t entity)
    {
        return GetLightWithoutCheckByEntity<vke_render::PointLight>(entity).positionWithRadius.w;
    }

    void VKE_INTEROP_CDECL SetPointLightRadius(uint32_t entity, float radius)
    {
        GetLightWithoutCheckByEntity<vke_render::PointLight>(entity).positionWithRadius.w = radius;
        MarkLightDirty<vke_render::PointLight>();
    }

    void VKE_INTEROP_CDECL GetSpotLightColor(uint32_t entity, Vector3<float> *color)
    {
        GetLightColor<vke_render::SpotLight>(entity, color);
    }

    void VKE_INTEROP_CDECL SetSpotLightColor(uint32_t entity, const Vector3<float> *color)
    {
        SetLightColor<vke_render::SpotLight>(entity, color);
    }

    float VKE_INTEROP_CDECL GetSpotLightIntensity(uint32_t entity)
    {
        return GetLightIntensity<vke_render::SpotLight>(entity);
    }

    void VKE_INTEROP_CDECL SetSpotLightIntensity(uint32_t entity, float intensity)
    {
        SetLightIntensity<vke_render::SpotLight>(entity, intensity);
    }

    float VKE_INTEROP_CDECL GetSpotLightRadius(uint32_t entity)
    {
        return GetLightWithoutCheckByEntity<vke_render::SpotLight>(entity).positionWithRadius.w;
    }

    void VKE_INTEROP_CDECL SetSpotLightRadius(uint32_t entity, float radius)
    {
        auto &light = GetLightWithoutCheckByEntity<vke_render::SpotLight>(entity);
        light.positionWithRadius.w = radius;
        UpdateSpotLight(entity, light);
    }

    float VKE_INTEROP_CDECL GetSpotLightInnerCone(uint32_t entity)
    {
        const float cosValue = GetLightWithoutCheckByEntity<vke_render::SpotLight>(entity).cone.x;
        return glm::acos(glm::clamp(cosValue, -1.0f, 1.0f));
    }

    void VKE_INTEROP_CDECL SetSpotLightInnerCone(uint32_t entity, float radians)
    {
        auto &light = GetLightWithoutCheckByEntity<vke_render::SpotLight>(entity);
        light.cone.x = glm::cos(radians);
        UpdateSpotLight(entity, light);
    }

    float VKE_INTEROP_CDECL GetSpotLightOuterCone(uint32_t entity)
    {
        const float cosValue = GetLightWithoutCheckByEntity<vke_render::SpotLight>(entity).cone.y;
        return glm::acos(glm::clamp(cosValue, -1.0f, 1.0f));
    }

    void VKE_INTEROP_CDECL SetSpotLightOuterCone(uint32_t entity, float radians)
    {
        auto &light = GetLightWithoutCheckByEntity<vke_render::SpotLight>(entity);
        light.cone.y = glm::cos(radians);
        UpdateSpotLight(entity, light);
    }
}
